File: Array/Python/Make Sum Divisible by P.py
# Problem link: https://leetcode.com/problems/make-sum-divisible-by-p/description
# Time: O(n)
# Space: O(n)
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        total = sum(nums)
        remainder = total % p

        if not remainder:
            return 0

        result, curr_sum = float('inf'), 0
        mapping = {0 : -1}

        for idx, num in enumerate(nums):
            curr_sum = (curr_sum + num) % p
            prefix = (curr_sum - remainder + p) % p
            if prefix in mapping:
                result = min(result, idx - mapping[prefix])
            mapping[curr_sum] = idx
        
        return -1 if result == len(nums) else result

File: Array/Java/Make Sum Divisible by P.java
// Problem link: https://leetcode.com/problems/make-sum-divisible-by-p/description
// Time: O(n)
// Space: O(n)
vclass Solution {
    public int minSubarray(int[] nums, int p) {
        int total = 0;

        for(int num : nums) {
            total = (total + num) % p;
        }

        if(total % p == 0) {
            return 0;
        }

        int curr_sum = 0, result = nums.length, remainder = total % p;
        HashMap<Integer, Integer> mapping = new HashMap<>();
        mapping.put(0, -1);

        for(int idx = 0; idx < nums.length; idx++) {
            curr_sum = (curr_sum + nums[idx]) % p;
            int prefix = (curr_sum - remainder + p) % p;

            if(mapping.containsKey(prefix)) {
                result = Math.min(result, idx - mapping.get(prefix));
            }

            mapping.put(curr_sum, idx);
        }

        return result == nums.length ? -1 : result;
    }
}
